raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -341,37 +356,59 @@ private VoteResponseData handleVoteRequest(
             return buildVoteResponse(Errors.INVALID_REQUEST, false);
         }
 
+        // we already confirmed that candidateEpoch >= quorum.epoch()
+        final boolean voteGranted;
         if (candidateEpoch > quorum.epoch()) {
-            // If the epoch is larger than our own, then we need to bump our own epoch,
-            // but we leave the vote undecided until we check the offset below
-            becomeUnattachedFollower(candidateEpoch);
-        }
+            final boolean isLeaderOrcandidate = quorum.isLeader() || quorum.isCandidate();
 
-        if (quorum.isLeader()) {
-            logger.debug("Ignoring vote request {} with epoch {} since we are already the leader in this epoch",
-                    request, candidateEpoch);
-            return buildVoteResponse(Errors.NONE, false);
-        } else if (quorum.isCandidate()) {
-            logger.debug("Ignoring vote request {} with epoch {} since we are already a candidate in this epoch",
-                    request, candidateEpoch);
-            return buildVoteResponse(Errors.NONE, false);
+            OffsetAndEpoch lastEpochEndOffsetAndEpoch = new OffsetAndEpoch(
+                    lastEpochEndOffset, lastEpoch);
+            voteGranted = lastEpochEndOffsetAndEpoch.compareTo(endOffset()) >= 0;
+
+            // even if we rejected the vote, we'd still need to bump our epoch,
+            // however, if the old state is leader or candidate, we should restart election right after the bump
+            if (!voteGranted) {
```

Comment:
Not really a big fan of slippered if-else. As we could see if we exit from this big if block the only logic to be checked against is `if(voteGranted)`. I would like to wait until we consolidate the if-else logic here a bit to make another concrete suggestion, but my take is that what could simplify the logic here is by putting the final return statements into here like:
```
 if (!voteGranted) {
     becomeUnattachedFollower(candidateEpoch);
     if (isLeaderOrcandidate) {
       becomeCandidate();
     }
  } else {
    becomeVotedFollower(candidateId, candidateEpoch);
  }
  return buildVoteResponse(Errors.NONE, voteGranted);
```
So that as a reader I understood this if logic ends here, which doesn't need to be traced down further through another big else block for logical consistency.